### PR TITLE
Return an intersection type when nameMock is used

### PR DESCRIPTION
**Pull request by martinssipenko**
This will return intersetion type when [`namedMock`](http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#named-mocks) function is used.

Currently when `Mockery::namedMock()` is used within tests PHPStan complains that the mocked object does not conform to what is expected.

```php
interface Bar {}

class Foo
{
    public function foo(Bar $bar) {
        $this->bar = $bar;
    }
}

// Works
$mock = Mockery::mock(Bar::class);
(new Foo)->foo($mock);

// Does not work
// Parameter #1 $bar of method Foo::foo() expects Bar, Mockery\MockInterface given. 
$namedMock = Mockery::namedMock('Buzz', Bar::class);
(new Foo)->foo($namedMock);
```

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Hi, this isn't sufficient as the extension will try to make a type from the first argument too.

**Comment by martinssipenko** [comment]
> Hi, this isn't sufficient as the extension will try to make a type from the first argument too.

Hm, I understand first argument also represents a type, essantually `namedMock('Foo', 'Bar')` is same as `calss Foo implements Bar`.

**Comment by ondrejmirtes** [comment]
From the documentation I don't understand what the method does: http://docs.mockery.io/en/latest/cookbook/class_constants.html

There are only classes involved, and in PHP you can't create intersection types from classes.

**Comment by martinssipenko** [comment]
When you mock something with `Mockery::mock(SomeInterface::class)` mockery creates a "fake" class that implements `SomeInterface` and a mock object from this "fake" class is created.

Now suppose you need to create two mock objects of same interface.
```
$mock1 = Mockery::mock(SomeInterface::class);
$mock2 = Mockery::mock(SomeInterface::class);
```

If you do `get_class($mock1)` and `get_class($mock2)` you will get the same result because the same "fake" class is used.

Using `namedMock` method you can get specify the name of the "fake" calss mockery creates, thus it allows to have two distinct objects that are created from different classes, but implement same interface.

**Comment by ondrejmirtes** [comment]
Yeah but if class name in the first argument isn't an existing one, we have to skip it when creating the intersection...

**Comment by martinssipenko** [comment]
Can you help me understand why? If we use `mock` method, then class/interface name (1st arg) can also be non-existant.

How do you think we can fix this?

**Comment by ondrejmirtes** [comment]
> then class/interface name (1st arg) can also be non-existant.

I don't think that's true.

**Comment by martinssipenko** [comment]
```php
<?php
// x.php

require_once __DIR__ . '/vendor/autoload.php';

$mock1 = Mockery::mock('WhaterverRandomNameIComeUpWith');
$mock2 = Mockery::mock('WhaterverRandomNameIComeUpWith');

var_dump(get_class($mock1));
var_dump(get_class($mock2));
```

```shell
php x.php 
/Users/msipenko/src/msipenko/forks/phpstan-mockery/x.php:8:
string(41) "Mockery_0__WhaterverRandomNameIComeUpWith"
/Users/msipenko/src/msipenko/forks/phpstan-mockery/x.php:9:
string(41) "Mockery_0__WhaterverRandomNameIComeUpWith"
```

**Comment by ondrejmirtes** [comment]
Yeah but PHPStan won't like it: https://phpstan.org/r/31d8e42c-5d2c-420a-b7d0-8161f271e686

Is it a valid use-case?

**Comment by martinssipenko** [comment]
Ah ok, I understand now, strangely I edited the code locally in my projects vendor directory, and just adding that entry to array  made PHPStan error go away.

Either way, what do you suggest we can do here?

**Comment by ondrejmirtes** [comment]
There has to be another extension similar to MockDynamicReturnTypeExtension but that skips the first parameter.

**Comment by martinssipenko** [comment]
Meybe we can just add something like this?

```php
		$args = $methodCall->args;
		if ($methodReflection->getName() === 'namedMock') {
			array_shift($args);
		}
```

**Comment by ondrejmirtes** [comment]
Nope, it'd be a mess, this needs a separate class.

File: src/Mockery/Type/MockDynamicNamedMockReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	public function getTypeFromStaticMethodCall(
+		MethodReflection $methodReflection,
+		StaticCall $methodCall,
+		Scope $scope
+	): Type
+	{
+		$defaultReturnType = new ObjectType('Mockery\\MockInterface');
+
+		$args = $methodCall->args;
+		array_shift($args);
```
Comment (ondrejmirtes) [inline src/Mockery/Type/MockDynamicNamedMockReturnTypeExtension.php]: You need to make sure the array has at least two elements.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Some code added to https://github.com/phpstan/phpstan-mockery/blob/master/tests/Mockery/MockeryTest.php to verify by running `vendor/bin/phing phpstan` this actually works would be nice.

**Comment by martinssipenko** [comment]
@ondrejmirtes, any updates on this?

**Comment by ondrejmirtes** [comment]
Merged and released, thanks.

**Comment by martinssipenko** [comment]
Thank you!

**Pull request merged**